### PR TITLE
SSO: Show default login form if we can not log user in with SSO

### PR DESCRIPTION
**Pull request by ebinnion**
Currently, if a user attempts to log in via SSO, and for some reason we can't find a user to login, we simply show a notice and the SSO login form again.

@rickybanister recently suggested that we show the default login form and then prompt the user (after logging in with Jetpack site credentials) to connect to WP.com. This is the first part of that flow.

To test:

- Checkout `update/sso-show-default-cant-find-user` branch
- Pick, or create, a test user on the remote site that is not connected to WP.com.
- And then either:
    - Make sure the test user has an email that doesn't match the WP.com user's email
    - In a compatibility plugin: `add_filter( 'jetpack_sso_match_by_email', '__return_false') ;`
- Go to `/wp-admin`
- Click "Log in with WordPress.com" button
- Approve the SSO process on WP.com
- When you land back on your site, you should see the username and password fields (aka default login form)

If you'd like to test more:

- In a compatibility plugin:
    - `add_filter( 'jetpack_remove_login_form', '__return_true' );`
- Follow same steps above to log in with WP.com
- When you land back on your Jetpack site, you should see the SSO login form
    - This is because we have disabled the default login form

cc @rickybanister for a design/flow review. @lezama and @enejb for code review.

### CODE REVIEW

**Comment by rickybanister** [comment]
The flow looks great—such an improvement to show the error with the regular log in form. This way users won't be able to fail repeatedly, which would be annoying.

File: modules/sso.php [hunk 1]
```diff
@@ -220,8 +220,16 @@ public function login_body_class( $classes ) {
...
+		/**
+		 * Should we show the default or SSO login form?
+		 *
+		 * $_GET['jetpack-sso-default-form'] is used to provide a fallback in case JavaScript is not enabled.
+		 *
+		 * The default_to_sso_login() method allows us to dynamically decide which type of login to show. The SSO module
+		 * uses the method to display the default login form if we can not find a user to log in via SSO. But, the method
+		 * could be filtered by a site admin to always show the default login form if that is preferred.
+		 */
+		if ( ! empty( $_GET['jetpack-sso-default-form'] )  || ! Jetpack_SSO_Helpers::default_to_sso_login() ) {
```
Comment (enejb) [inline modules/sso.php]: Thanks for the comment above. 
I find this really hard to follow. I don't know if returning true here will display the default form or display the sso form. 
My guess  is that if the statment is true that we should show the default form. 

I think to make it more clear we could do something like
```
$show_default_wp_login_form = ( isset( $_GET['jetpack-sso-default-form'] ) || ! Jetpack_SSO_Helpers::display_sso_login() )
if ( $show_default_wp_login_form ) {
return $classes;
}
```

Comment (ebinnion) [inline modules/sso.php]: Thanks for the feedback.

I like your option.

Or, perhaps we just updated the first part of the comment to be:

```
Should we show the default login form?
```

**Comment by ebinnion** [comment]
I've updated the logic that we use for deciding whether to display the SSO login form or not.

So, you'll want to test with the above instructions. But, you should also test that our fallback for no js works.

In Chrome, I use a plugin called "Quick Javascript Switcher". To test that:

- Turn off JS
- Go to `$site/wp-admin`
- Click "Log in with username and password"
- Click "Log in with WordPress.com"
- Ensure that both work

File: modules/sso.php [hunk 2]
```diff
@@ -808,6 +815,8 @@ function handle_login() {
 			exit;
 		}
 
+		add_filter( 'jetpack_sso_default_to_sso_login', '__return_false' );
```
Comment (enejb) [inline modules/sso.php]: Should we update the filter as well. I think its fine.

Comment (ebinnion) [inline modules/sso.php]: I think it's ok to leave the filter as is.

**Comment by jeherve** [comment]
This seems to work in my tests.

However, it doesn't seem to be compatible with Jetpack's Protect module. When I'm redirected to the log in form, it includes the math fallback:

screen shot 2016-06-30 at 3 15 50 pm

**Comment by ebinnion** [comment]
@jeherve - If you check out `branch-4.0` and then go to the login page, do you get the protect math? I did in my testing. When I looked a bit into the logic, I saw this:

> The math captcha fallback if we can't talk to the Protect API

https://github.com/Automattic/jetpack/blob/127f0fe6778eb14af3802ecef51f8f52c4fb9d72/modules/protect/math-fallback.php#L5

I'm not very acquainted with the protect module yet, but it seems like there may be a communication error to the Protect API. Perhaps @samhotchkiss can chime in, but I don't think you seeing the protect math is related to this PR.

**Comment by lezama** [comment]
🚢

**Comment by dereksmart** [comment]
The only two reasons that a user would see math are: 

- Unable to communicate with the bruteprotect API
- The user is coming from an IP that has been "soft blocked".  This means that there have been enough failed logins within a certain amount of time to trigger a semi-lockout.  More failed login attempts would result in a "hard block", and waiting several minutes would result in the "soft block" to clear.

**Comment by ebinnion** [comment]
I have cherry-picked the commit into `branch-4.1`.
